guide/src/free-threading.md

```diff
@@ -0,0 +1,153 @@
+for porting C extensions, and [PEP 703](https://peps.python.org/pep-0703/),
+which provides the technical background for the free-threading implementation in
+CPython.
+
+The main benefit for supporting free-threaded Python is that both pure Python
+code and code interacting with Python via the CPython C API can run
+simultaneously on multiple OS threads.  This means it's much more
+straightforward to achieve multithreaded parallelism in the Python layer, using
+the `threading` module, without any need to ensure the GIL is released during
+compute-heavy tasks, or any worries about [Amdahl's
+law](https://en.wikipedia.org/wiki/Amdahl%27s_law) limiting parallel
+performance. If you have ever needed to use `multiprocessing` to achieve a
+speedup for some Python code, free-threading will likely allow the use of Python
+threads instead for the same workflow.
+
+PyO3's support for free-threaded Python will enable authoring native Python
+extensions that are thread-safe by construction, with much stronger safety
+guarantees than C extensions. Our goal is to enable ["fearless
+concurrency"](https://doc.rust-lang.org/book/ch16-00-concurrency.html) in the
+native Python runtime by building on the rust `Send` and `Sync` traits.
+
+This document provides advice for porting rust code using PyO3 to run under
+free-threaded Python. While many simple PyO3 uses, like defining an immutable
+Python class, will likely work "out of the box", there are currently some
+limitations.
+
+## Many symbols exposed by PyO3 have `GIL` in the name
+
+We are aware that there are some naming issues in the PyO3 API that make it
+awkward to think about a runtime environment where there is no GIL. We plan to
+change the names of these types to de-emphasize the role of the GIL in future
+versions of PyO3, but for now you should remember that the use of the term `GIL`
+in functions and types like `with_gil` and `GILOnceCell` is historical.
+
+Instead, you can think about whether or not a rust thread is attached to a
+Python interpreter runtime. See [PEP
+703](https://peps.python.org/pep-0703/#thread-states) for more background about
+how threads can be attached and detached from the interpreter runtime, in a
+manner analagous to releasing and acquiring the GIL in the GIL-enabled build.
+
+Calling into the CPython C API is only legal when an OS thread is explicitly
+attached to the interpreter runtime. In the GIL-enabled build, this happens when
+the GIL is acquired. In the free-threaded build there is no GIL, but the same C
+macros that release or acquire the GIL in the GIL-enabled build instead ask the
+interpreter to attach the thread to the Python runtime, and there can be many
+threads simultaneously attached.
+
+The main reason for attaching to the Python runtime is to interact with Python
+objects or call into the CPython C API. To interact with the Python runtime, the
+thread must register itself by attaching to the interpreter runtime.
```

Comment:
Maybe mention `with_gil` here as the way to attach and the plan is that it will continue to exist under a new name?